topo/topo.go

```diff
@@ -522,3 +543,83 @@ func DeleteTopology(ctx context.Context, params TopologyParams) error {
 
 	return nil
 }
+
+// serviceToProto creates the service mapping.
+func serviceToProto(s *corev1.Service, m map[uint32]*tpb.Service) error {
+	if s == nil || m == nil {
+		return fmt.Errorf("service and map must not be nil")
+	}
+	if len(s.Status.LoadBalancer.Ingress) == 0 {
+		return fmt.Errorf("service %s has no external loadbalancer configured", s.Name)
+	}
+	for _, p := range s.Spec.Ports {
+		k := uint32(p.Port)
+		service, ok := m[k]
+		if !ok {
+			service = &tpb.Service{
+				Name:   p.Name,
+				Inside: k,
+			}
+			m[k] = service
+		}
+		if service.Name == "" {
+			service.Name = p.Name
+		}
+		service.Outside = uint32(p.TargetPort.IntVal)
+		service.NodePort = uint32(p.NodePort)
+		service.InsideIp = s.Spec.ClusterIP
+		service.OutsideIp = s.Status.LoadBalancer.Ingress[0].IP
+	}
+	return nil
+}
+
+// GetTopologyServices returns the topology information.
+func GetTopologyServices(ctx context.Context, params TopologyParams) (*tpb.Topology, error) {
+	topopb, err := Load(params.TopoName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load %s: %+v", params.TopoName, err)
+	}
+
+	var t resourcer
+	if params.fakeTopo != nil {
+		t = params.fakeTopo
+	} else {
+		t, err = New(params.Kubecfg, topopb, params.TopoNewOptions...)
```

Comment:
Maybe a solution would be to stub New which should return a resourcer

Then each test case provides a stub for the New func which returns a fakeTopo

I think this is the cleanest solution, but open to discussion